src/uct/ib/rdmacm/rdmacm_iface.c

```diff
@@ -158,53 +166,78 @@ static void uct_rdmacm_iface_process_conn_req(uct_rdmacm_iface_t *iface,
     rdma_destroy_id(event->id);
 }
 
-static void uct_rdmacm_iface_process_event(uct_rdmacm_iface_t *iface, struct rdma_cm_event *event)
+static void uct_rdmacm_iface_return_cm_id(uct_rdmacm_iface_t *iface,
+                                          uct_rdmacm_cm_id_t *rdmacm_cm_id)
+{
+    ucs_debug("destroying cm_id %p", rdmacm_cm_id->cm_id);
+
+    ucs_list_del(&rdmacm_cm_id->list);
+    if (rdmacm_cm_id->ep != UCT_RDMACM_IFACE_BLOCKED_NO_EP) {
+        rdmacm_cm_id->ep->is_on_cm_ids_list = 0;
+    }
+    rdma_destroy_id(rdmacm_cm_id->cm_id);
+    ucs_free(rdmacm_cm_id);
+    iface->num_cm_id_in_quota++;
+}
+
+static int uct_rdmacm_iface_process_event(uct_rdmacm_iface_t *iface, struct rdma_cm_event *event)
 {
     struct sockaddr *remote_addr = rdma_get_peer_addr(event->id);
     uct_rdmacm_md_t *rdmacm_md   = (uct_rdmacm_md_t *)iface->super.md;
     char ip_port_str[UCS_SOCKADDR_STRING_LEN];
     uct_rdmacm_priv_data_hdr_t *hdr;
     struct rdma_conn_param conn_param;
+    uct_rdmacm_cm_id_t *rdmacm_cm_id;
+    uct_rdmacm_ep_t *ep = NULL;
+    int destroy_cm_id = 0;
 
-    ucs_debug("rdmacm event (fd=%d) on %s: %s. Peer: %s.",
-              iface->event_ch->fd, (iface->is_server ? "server" : "client"),
+    ucs_debug("rdmacm event (fd=%d cm_id %p) on %s: %s. Peer: %s.",
+              iface->event_ch->fd, event->id, (iface->is_server ? "server" : "client"),
               rdma_event_str(event->event),
               ucs_sockaddr_str(remote_addr, ip_port_str, UCS_SOCKADDR_STRING_LEN));
 
-    ucs_assert((iface->cm_id == event->id) ||
-               ((event->event == RDMA_CM_EVENT_CONNECT_REQUEST) && (iface->cm_id == event->listen_id)));
+    if (iface->is_server) {
+        ucs_assert((iface->cm_id == event->id) ||
+                   ((event->event == RDMA_CM_EVENT_CONNECT_REQUEST) &&
+                    (iface->cm_id == event->listen_id)));
+    } else {
+        rdmacm_cm_id = (uct_rdmacm_cm_id_t *)event->id->context;
+        ep = rdmacm_cm_id->ep;
+    }
 
     /* The following applies for rdma_cm_id of type RDMA_PS_UDP only */
     switch (event->event) {
     case RDMA_CM_EVENT_ADDR_RESOLVED:
         /* Client - resolve the route to the server */
-        ucs_assert(iface->ep != NULL);
-        if (iface->ep == UCT_RDMACM_IFACE_BLOCKED_NO_EP) {
-            uct_rdmacm_iface_client_start_next_ep(iface);
+        if (ep == UCT_RDMACM_IFACE_BLOCKED_NO_EP) {
```

Comment:
please add log which will show the event type

Reply:
you mean in addition to the one at the begining of the function?

Reply:
can be anywhere, as long as it's clear which event has arrived and on which cm_id/ep